### PR TITLE
Stream test changes

### PR DESCRIPTION
**Pull request by alees24**
1. Slightly extends the functionality of the existing USB streaming code to support performance monitoring/unidirectional streams better, allowing the host application to adapt automatically to the test being performed.

2. Migrates the streaming code into the lib/testing directory alongside the other usb_testutils code, importantly allowing this streaming test to be used in multiple top-level tests in various ways, such as:

- intermixing Isochronous streams with Bulk Transfer streams
- suspending/resuming/sleep /whilst/ actively transferring data
- system integration testing; top-level tests may perform USB traffic whilst exercising other blocks

**Please Note** Most of this commit is just separating some of the code from tests/usbdev_stream_test.c out into lib/testing/usb_testutils_streams.c/h but sufficient editing is required in this refactoring that the diff doesn't make that very clear.

3. Renames the functions, data types and macros to suit its new location and reduce the chances of name conflicts.

### CODE REVIEW

File: sw/device/tests/usbdev_stream_test.c
Comment (marnovandermaas) [inline sw/device/tests/usbdev_stream_test.c]: Which part of the byte is the number and which are flags?

File: sw/device/tests/usbdev_stream_test.c
Comment (marnovandermaas) [inline sw/device/tests/usbdev_stream_test.c:247]: Maybe change the commit message to renaming and refactoring.

**Review by marnovandermaas — APPROVED** [review]
@alees24 very nice PR! I appreciate the work in splitting the code into separate commits.
I have not tested this PR, but feel it is ok to approve it because it is a software change and mostly refactoring.
Nevertheless, it would be good to get another set of eyes on this before merging.

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (engdoreis) [inline sw/device/lib/testing/usb_testutils_streams.c:240]: Consider using an enum, a `#define`, or keep as a variable but follow naming conversion to define this constant as per the [style guide](https://opentitan.org/book/doc/contributing/style_guides/c_cpp_coding_style.html#constants-and-preprocessor-macros).

Comment (engdoreis) [inline sw/device/lib/testing/usb_testutils_streams.c:240]: Consider using an enum, a `#define`, or keep as a variable but follow naming conversion to define this constant as per the [style guide](https://opentitan.org/book/doc/contributing/style_guides/c_cpp_coding_style.html#constants-and-preprocessor-macros).

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (engdoreis) [inline sw/device/lib/testing/usb_testutils_streams.c:243]: I could not find this function definition. I think you forgot to upload it, as this code is not compiled in CI.

Comment (engdoreis) [reply sw/device/lib/testing/usb_testutils_streams.c:243]: You can probably use `mmio_region_memcpy_from_mmio32` here.

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (engdoreis) [inline sw/device/lib/testing/usb_testutils_streams.c:343]: Could you declare the handlers types in the header to improve readability?
i.e 
```c
typedef void (*usb_testutils_rx_handler)(void *, dif_usbdev_rx_packet_info_t, dif_usbdev_buffer_t);
```

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (engdoreis) [inline sw/device/lib/testing/usb_testutils_streams.c:395]: As per the [style guide](https://opentitan.org/book/doc/contributing/style_guides/c_cpp_coding_style.html#formatting-of-loops-and-conditionals), single line conditionals and loops must use braces.

**Review by engdoreis — COMMENTED** [review]
Thanks @alees24 for the PR.
More testutils are always good, and as part of the [manufacturing](https://github.com/orgs/lowRISC/projects/13/views/40) effort, we are refactoring the testutils to never abort and return `status_t` instead (See details in this issue https://github.com/lowRISC/opentitan/issues/17495). 

So could you change your new testutils functions return `status_t` and replace the macros `CHECK`, `CHECK_DIF_OK` by `TRY_CHECK` and `TRY` respectively?
You can find some examples in the testutils which were refactored already, i.e. [flash_ctrl](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/testing/flash_ctrl_testutils.c).

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (alees24) [reply sw/device/lib/testing/usb_testutils_streams.c:243]: I haven't forgotten that function; it's in a separate PR on another branch, but it's inaccessible in this PR because the _MEM_FASTER macro is set to 0.

What that routine does is to introduce replacement routines for mmio_memcpy_from|to_mmio32 for the usbdev code because the generic function is very slow, but it also induces functional failures in DV chip sim so the replacement function is necessary. For now I switch it in manually, but really we need the replacement routines to be ratified and become the default.

The problems arise from the packet buffer memory being mapped as if it were CSRs, and it is indeed only word-addressable in the same fashion, but its behaviour differs because it must handle much more data in a performant manner, and there is no reset logic.

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (alees24) [reply sw/device/lib/testing/usb_testutils_streams.c:342]: Thanks for the suggestion; I do intend to do that later, along with various other non-functional tidyups throughout usb_testutils.

I have made the status_t/TRY changes to the newly-added usb_testutils functions, but left the existing usb_testutils code unmodified by the same token; TBD in a subsequent PR

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (alees24) [reply sw/device/lib/testing/usb_testutils_streams.c]: I'm somewhat surprised that clang_format_fix doesn't modify that, but thank you for catching it.

**Review by a-will — APPROVED** [review]

File: sw/device/tests/usbdev_stream_test.c
Comment (marnovandermaas) [reply sw/device/tests/usbdev_stream_test.c:159]: We talked about this in person. I think this is fine for now.

File: sw/device/tests/usbdev_stream_test.c
Comment (marnovandermaas) [reply sw/device/tests/usbdev_stream_test.c]: This is done.

File: sw/device/tests/usbdev_stream_test.c
Comment (marnovandermaas) [reply sw/device/tests/usbdev_stream_test.c]: This was added in a later commit.

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (alees24) [reply sw/device/lib/testing/usb_testutils_streams.c:240]: It has that name because the code and naming convention matches that in the DIF implementation of this routine. This has now been superseded by an effort to implement the faster routines in the DIF itself, so I've removed the code from here.

**Comment by marnovandermaas** [comment]
> I have been trying hard to sequence things, but it becomes tricky with a lot of branches/PRs and their dependencies. I had thought it would be okay to leave the function implementation for a subsequent PR since its invocation is inaccessible in this PR.
> 
> Having tried to migrate the faster routines into a DIF I've bumped into other environment/software requirements anyway, so I'll just remove the '...Faster' method code for now and leave this as a TODO. Hope that's okay.

Yes, I believe removing this `USBUTILS_MEM_FASTER` block would be best and then we can merge this PR.

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (marnovandermaas) [inline sw/device/lib/testing/usb_testutils_streams.c:241]: Would you mind squashing this change into the first commit into this PR?

**Review by marnovandermaas — COMMENTED** [review]
Thanks removing the unused code, would you mind moving that change to the first commit in this PR instead of the last?

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (engdoreis) [inline sw/device/lib/testing/usb_testutils_streams.c]: Sorry for don't noticing this before, but in the first commit this enum is being defined in the file `sw/device/tests/usbdev_stream_test.c`, so could you move this definition to a common place? Maybe to the `usb_testutils.h`? I fine if do it in a follow-up PR.

File: sw/device/lib/testing/usb_testutils_streams.c
Comment (engdoreis) [reply sw/device/lib/testing/usb_testutils_streams.c:241]: Yah, the later commits change the previous one, so I also think they should be squashed.

**Comment by marnovandermaas** [comment]
Thanks @alees24 Let's wait for CI to finish and then I'll merge this.

**Pull request merged**
